### ISSUE TITLE
ExceptionHandler inconsistently handling Exceptions in Maybe objects

### ISSUE THREAD

**Issue by markjfisher** [micronaut-projects/micronaut-core#1594]
- [x] Steps to reproduce provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Write an ExceptionHandler that handles Exception.class
2. Create an error in controller that returns a Maybe 
3. Sometimes see the handler kick in (requires JVM restart if it doesn't handle it)

### Expected Behaviour

A generic Exception should be handled by the ExceptionHandler consistently

### Actual Behaviour

Sometimes does, sometimes doesn't. Non deterministic. Have to restart JVM to fix, and doesn't always. We see this in K8S environments with multiple pods, some pods work, some don't.

### Environment Information

- **Operating System**: Linux based tested
- **Micronaut Version:** 1.1.0
- **JDK Version:** jdk-11

### Example Application

https://github.com/markjfisher/mn-maybe-test

This was discussed in https://kotlinlang.slack.com/archives/CDQF87V4N/p1555512486040900

**Comment by jameskleeh** [comment]
The behavior of your exception handler being chosen is what is incorrect. There is an exception handler already registered for a more specific exception (`JsonExceptionHandler`). If you don't want that handler to be triggered, then your handler should replace it `@Replaces(JsonExceptionHandler.class)`.

**Comment by markjfisher** [comment]
But this doesn't explain why the issue is non deterministic and requires a JVM restart, that sometimes still fails.
Also, we're not going near Json, so I don't see what that has to do with it.
Finally, the blockingGet() on the maybe completely fixes the issue (but is obviously not usable), so why does that work ok?

**Comment by jameskleeh** [comment]
> Also, we're not going near Json, so I don't see what that has to do with it.

That isn't true. The exception being thrown is a `JsonParseException` triggered by the data service:

```
    private val mapper = jacksonObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)

    fun createList(num: Int): List<Data> {
        val data = mapper.readValue<Data>("bad data")
        return MutableList(num) { data }
    }
```

> Finally, the blockingGet() on the maybe completely fixes the issue

I haven't dug into the success case so I'm not sure why. Could be a different problem

**Comment by markjfisher** [comment]
oh, yes, sorry we are doing json here. but yes, the problem for us is that on some JVM instances it works, others it doesn't (not that I stumbled across blockingGet fixing it).

I can clearly fix the overriding of JsonExceptionHandler, but shouldn't Micronaut be able to determine correctly the handler to choose (which seems to be the real issue) from the available ones, and an Exception.class handler would be the ultimate "catch all"?

Is there an easy way to get the Handlers that are registered at runtime so we can determine which ones we need to replace, as this wasn't obvious to us?

Many thanks for your help on this.

**Comment by jameskleeh** [comment]
>  but shouldn't Micronaut be able to determine correctly the handler to choose (which seems to be the real issue) from the available ones

Yes it should and like I said what you are considering the error condition is actually the correct behavior. A json exception is being thrown and thus the json exception handler is being used.

> an Exception.class handler would be the ultimate "catch all"?

You are considering the `Exception` class handler to be of the highest priority, but in fact it is the lowest priority. If any of the existing exception handlers match the exception being thrown, they will be used before a global "catch all"

> Is there an easy way to get the Handlers that are registered at runtime so we can determine which ones we need to replace, as this wasn't obvious to us?

Unfortunately there isn't an easy way to determine which ones will be registered at runtime, however I think if you just look for classes that implement ExceptionHandler in your IDE, you will see there are not very many and most of them you would likely not want to override.

**Comment by markjfisher** [comment]
I'd consider a "catch all" to be the lowest priority too, otherwise wouldn't be able to write other handlers, so in fact, and came to realise (as you say) that what I'm considering the correct case is the wrong case!

Ok, we'll implement the change suggested, and leave you with the bug of why it's picking Exception.class handler at all :)

Thanks for the explanation, it also clarifies why our specific handlers work solidly.

**Comment by markjfisher** [comment]
Great stuff, thanks!

**Comment by alexwilbur** [comment]
@jameskleeh is this definitely fixed? I am trying 1.1.2 (I think this was merged into 1.1.1) and I still see the same issue - non deterministic selection of the exception mapper.

I'm using AWS Lambda. I made a concrete exception class which is derived from RuntimeException and then implemented a concrete exception handler and an additional exception handler for RuntimeException.

I then throw the concrete exception inside an HTTP GET request handler.

Apparently, at  "random" either my intended concrete exception mapper is invoked (happy path), OR a generic exception mapper for RuntimeException is invoked. If I cold start the Lambda the behavior is just unpredictable.

What might govern the decision making about which exception handler to invoke?
